### ISSUE TITLE
lit-html type checking doesn't detect unmatched closing tag

### ISSUE THREAD

**Issue by ZufengW** [runem/lit-analyzer#54]
If I have an unclosed opening tag it detects the error correctly.
```
    return html`
      <div>
    `;
```
`error TS2322: [lit] This tag isn't closed.`

But if I have an unmatched _closing tag_, no error is reported.
```
    return html`
      </div>
    `;
```
My VSCode lit-plugin extension doesn't detect it. But it would be something good to warn for.

**Comment by runem** [comment]
Hi, thanks for your suggestion!

I use `parse5` to parse HTML. However, it seems like `parse5` doesn't emit any nodes for your second example. In order to fix it I'll have to use another parser. I have also been thinking of looking for another parser because of `parse5` not seeing the different between lowercase/uppercase in attribute names. 

I fear it would be a large task to replace the parser, so unfortunately I wouldn't have time to do it in the near future. I'll keep you updated here :-)

**Comment by justinfagnani** [comment]
@rune you can get the original-cased attribute name with parse5 by using the source location of the attribute.

I think you can also infer a missing end node from source locations too. I'll try when I'm back at a computer...

**Comment by runem** [comment]
This is the result I get when parsing "`</div>`" (I could be missing some parse5 options that could make it possible, though):
```js
{ nodeName: '#document-fragment', childNodes: [] }
```

Regarding original-cased attribute names, my challenge actually goes deeper than just the handling of casing. Parse5 focuses on being spec compliant, so they won't accept PRs for any features that violate the spec or provide various "relaxed" parsing modes. I also think that's why the above example results in zero nodes. 

My main challenge regarding this is to detect multiple usages of the same name. For example, `<my-element .myValue=${1}" .myValue=${2} .myValue=${3}></my-element>` would be parsed as the following. It would only report the value of the first occurrence, and only the source location of the last occurrence.

```js
{ nodeName: '#document-fragment',
  childNodes: [ 
    { 
       nodeName: 'my-element',
       tagName: 'my-element',
       attrs: [ { name: '.myvalue', value: '____' } ],
       ...,
       sourceCodeLocation: { 
          attrs: {
             myvalue: { 
                startCol: 41,
                endCol: 53,
                ... 
             }
          },
          ...
       }
    }
  ]
}
```

**Comment by 43081j** [comment]
When experimenting with this problem a while back, again with parse5, I reached the conclusion that you'd just have to first detect the presence of the attribute then check the original source (using the source locations of the start tag) for duplicates.

parse5 does lowercase and de-dupe because thats what it should do as a spec-compliant parse afaik. 

similar would have to be done to detect casing i imagine but that would be much easier i suspect (getting the raw value at a location in the source).
